### PR TITLE
SILGen: Fix setting variables with property wrappers marked objc dynamic

### PR DESCRIPTION
**Pull request by aschwaighofer**
rdar://54181647

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/SILGen/SILGenLValue.cpp [hunk 1]
```diff
@@ -1430,7 +1430,14 @@ namespace {
 
         // Create the allocating setter function. It captures the base address.
         auto setterInfo = SGF.getConstantInfo(setter);
-        SILValue setterFRef = SGF.emitGlobalFunctionRef(loc, setter, setterInfo);
+        SILValue setterFRef;
+        if (setter.hasDecl() && setter.getDecl()->isObjCDynamic()) {
```
Comment (slavapestov) [inline lib/SILGen/SILGenLValue.cpp]: Can the setter be a class method or a witness method also

Comment (aschwaighofer) [reply lib/SILGen/SILGenLValue.cpp]: I don't think so, this code is for property wrapped vars.

This:

class SomeWrapperTests {                                                        
  @SomeWrapper var someWrapper: Int = 0   

will just emit a function ref.

I am guessing we treat this like stored properties.

Comment (jrose-apple) [reply lib/SILGen/SILGenLValue.cpp]: Property-wrapped vars should definitely be overridable in subclasses.

Comment (aschwaighofer) [reply lib/SILGen/SILGenLValue.cpp]: Hmm, then the code before this patch is broken:
```
 @propertyWrapper
 public struct SomeWrapper {
   private var value: Int

 
    public init(wrappedValue: Int) {
     value = wrappedValue
   }

 
    public var wrappedValue: Int {
     get { value }
     set { value = newValue }
   }
 }
```

SILGen will emit a direct function_ref and not a class_method call for the setter in the example below:

```
  class SomeWrapperTests {
   @SomeWrapper var someWrapper: Int = 0
   func testAssignment() {
     someWrapper = 1000
   }
 }
```


```
sil hidden [ossa] @$s7Wrapper04SomeA5TestsC14testAssignmentyyF : $@convention(method) (@guaranteed SomeWrapperTests) -> () {
// %0                                             // users: %12, %6, %1
bb0(%0 : @guaranteed $SomeWrapperTests):
  ...
  %9 = function_ref @$s7Wrapper04SomeA0V12wrappedValueACSi_tcfCTc : $@convention(thin) (@thin SomeWrapper.Type) -> @owned @callee_guaranteed (Int) -> SomeWrapper
  %10 = apply %9(%8) : $@convention(thin) (@thin SomeWrapper.Type) -> @owned @callee_guaranteed (Int) -> SomeWrapper
  // function_ref SomeWrapperTests.someWrapper.setter
  %11 = function_ref @$s7Wrapper04SomeA5TestsC04someA0Sivs : $@convention(method) (Int, @guaranteed SomeWrapperTests) -> ()
  %12 = copy_value %0 : $SomeWrapperTests
  %13 = partial_apply [callee_guaranteed] %11(%12) : $@convention(method) (Int, @guaranteed SomeWrapperTests) -> ()
  assign_by_wrapper %5 : $Int to %7 : $*SomeWrapper, init %10 : $@callee_guaranteed (Int) -> SomeWrapper, set %13 : $@callee_guaranteed (Int) -> ()
  end_access %7 : $*SomeWrapper 
 
```

Comment (jrose-apple) [reply lib/SILGen/SILGenLValue.cpp]: Yikes! We should fix that too, then. :-) I guess I'll file a separate Radar.

Comment (jrose-apple) [reply lib/SILGen/SILGenLValue.cpp]: Filed [SR-11571](https://bugs.swift.org/browse/SR-11571), soon to be cloned.

**Pull request merged**